src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpSpanMappingService.cs

```diff
@@ -0,0 +1,125 @@
+            ITextSnapshot textSnapshot)
+        {
+            if (lspDocumentMappingProvider is null)
+            {
+                throw new ArgumentNullException(nameof(lspDocumentMappingProvider));
+            }
+
+            if (textSnapshot == null)
+            {
+                throw new ArgumentNullException(nameof(textSnapshot));
+            }
+
+            if (documentSnapshot is null)
+            {
+                throw new ArgumentNullException(nameof(documentSnapshot));
+            }
+
+            _lspDocumentMappingProvider = lspDocumentMappingProvider;
+
+            _textSnapshot = textSnapshot;
+            _documentSnapshot = documentSnapshot;
+        }
+
+        public async Task<ImmutableArray<RazorMappedSpanResult>> MapSpansAsync(
+            Document document,
+            IEnumerable<TextSpan> spans,
+            CancellationToken cancellationToken)
+        {
+            return await MapSpansAsync(document, spans, _textSnapshot.AsText(), _documentSnapshot.Snapshot.AsText(), cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<ImmutableArray<RazorMappedSpanResult>> MapSpansAsync(
+            Document document,
+            IEnumerable<TextSpan> spans,
+            SourceText sourceTextGenerated,
+            SourceText sourceTextRazor,
+            CancellationToken cancellationToken)
+        {
+            if (spans == null)
+            {
+                throw new ArgumentNullException(nameof(spans));
+            }
+
+            var projectedRanges = spans.Select(span => span.AsLSPRange(sourceTextGenerated)).ToArray();
+
+            var mappedResult = await _lspDocumentMappingProvider.MapToDocumentRangesAsync(
+                RazorLanguageKind.CSharp,
+                _documentSnapshot.Uri,
+                projectedRanges,
+                cancellationToken).ConfigureAwait(false);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var results = ImmutableArray.CreateBuilder<RazorMappedSpanResult>();
+
+            if (mappedResult is null)
+            {
+                return results.ToImmutable();
+            }
+
+            foreach (var mappedRange in mappedResult.Ranges)
+            {
+                var mappedSpan = mappedRange.AsTextSpan(sourceTextRazor);
+                var linePositionSpan = sourceTextRazor.Lines.GetLinePositionSpan(mappedSpan);
+                var filePath = _documentSnapshot.Uri.LocalPath;
+                results.Add(new RazorMappedSpanResult(filePath, linePositionSpan, mappedSpan));
+            }
+
+            return results.ToImmutable();
+        }
+
+        // Internal for testing use only
+        internal async Task<IEnumerable<(string filePath, LinePositionSpan linePositionSpan, TextSpan span)>> MapSpansAsyncTest(
+            IEnumerable<TextSpan> spans,
+            SourceText sourceTextGenerated,
+            SourceText sourceTextRazor)
+        {
```

Comment:
SourceText provided by the test suite to get around inability to mock static extensions (`TextSpan.AsText()`).

Conversion from `RazorMappedSpanResult` to tuple to get around dependency/protection issues similar to what we do for the Razor Span/Excerpt Services.